### PR TITLE
Introduce HttpEngine

### PR DESCRIPTION
**Pull request by martinbonnin**
Introduce a lower level HttpEngine interface on top of OkHttp and NSURLSession. This allows to factor the error handling code and also the code that prepares the requests. This might make multiplatform upload work.

### CODE REVIEW

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpUrl.kt [hunk 1]
```diff
@@ -0,0 +1,5 @@
+package com.apollographql.apollo3.network.http
+
+expect class HttpUrlBuilder(baseUrl: String, queryParameters: Map<String, String>) {
+  fun build(): String
+}
```
Comment (sav007) [inline apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpUrl.kt:5]: is there plans to extend this class in future? It looks a little odd to have class with one fun that returns just a string? Should it be just top level function `buildUrl`? Or  even better introduce `Url / Uri` class that represents url and will have companion factory object?

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt [hunk 2]
```diff
@@ -129,27 +131,21 @@ class ApolloClient private constructor(
       )
     }
 
-    /**
-     * internal because only used from tests
-     */
-    internal fun interceptors(interceptors: List<ApolloRequestInterceptor>) = apply {
```
Comment (sav007) [inline apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt:135]: why we making these API non internal? Are we using it in different module now?

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/ApolloHttpNetworkTransport.kt [hunk 3]
```diff
@@ -21,4 +35,98 @@ expect class ApolloHttpNetworkTransport constructor(
      * - on Android, it is used to set  [OkHttpClient.readTimeout]
      */
     readTimeoutMillis: Long = 60_000,
-) : NetworkTransport
+
+    val engine: HttpEngine = DefaultHttpEngine(connectTimeoutMillis, readTimeoutMillis)
```
Comment (sav007) [inline apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/ApolloHttpNetworkTransport.kt:39]: all private `val`s?

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpEngine.kt [hunk 4]
```diff
@@ -0,0 +1,56 @@
+package com.apollographql.apollo3.network.http
+
+import okio.BufferedSink
+import okio.BufferedSource
+
+/**
+ * A wrapper around platform specific engines
+ */
+interface HttpEngine {
```
Comment (sav007) [inline apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpEngine.kt:9]: Ktor effect :)

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpUrl.kt (hunk 1 above)
Comment (martinbonnin) [reply apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/HttpUrl.kt:5]: I'd love the Kotlin stlib to ship something there so we can remove this. I'm a bit hesitant on introducing a MPP `Url/Uri` as `Url` pulls a lot of potentially complex things (scheme, hosts, absolute vs relative, etc..) hence the very targetted API. Agreed there's no need for a class though, I'll replace with `buildUrl`

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt (hunk 2 above)
Comment (martinbonnin) [reply apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/ApolloClient.kt:135]: We were using it from `apollo-integration-kotlin` but it wasn't strictly need as the public API could do too, I put everything back to `internal`

File: apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/ApolloHttpNetworkTransport.kt (hunk 3 above)
Comment (martinbonnin) [reply apollo-runtime-kotlin/src/commonMain/kotlin/com/apollographql/apollo3/network/http/ApolloHttpNetworkTransport.kt]: Yup!

**Review by sav007 — APPROVED** [review]

**Pull request merged**